File: fix_codeblocks.py
#!/usr/bin/env python3
"""
Automatically fix code blocks missing language tags in MDX files.
"""
import re
import sys
from pathlib import Path

def determine_language(content, prev_lines, next_lines):
    """Determine the appropriate language tag based on content and context."""
    content_lower = content.lower()
    prev_text = '\n'.join(prev_lines[-5:]).lower()
    next_text = '\n'.join(next_lines[:3]).lower()
    
    # HTTP endpoints
    if re.search(r'^(GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+/', content, re.MULTILINE):
        return 'http'
    
    # curl commands
    if re.search(r'^\s*curl\s', content, re.MULTILINE):
        return 'bash'
    
    # YAML/Docker Compose
    if re.search(r'^(services:|version:|networks:|volumes:|image:|ports:|environment:)', content, re.MULTILINE):
        return 'yaml'
    
    # Caddyfile
    if re.search(r'^\w+\.(com|dev|local|net)\s*{', content, re.MULTILINE) or 'caddy' in prev_text:
        return 'caddy'
    
    # JSON
    if (content.strip().startswith('{') or content.strip().startswith('[')) and re.search(r'["\']:\s*["\'{[]', content):
        return 'json'
    
    # XML
    if re.search(r'<\?xml|<roblox|<Item|<Properties|</\w+>', content):
        return 'xml'
    
    # Lua
    if re.search(r'(function|local|if\s+.+\s+then|end\b|return\b)', content, re.MULTILINE):
        return 'lua'
    
    # Dockerfile
    if re.search(r'^(FROM|RUN|COPY|WORKDIR|ENTRYPOINT|CMD|EXPOSE|ENV|ARG)\s', content, re.MULTILINE):
        return 'dockerfile'
    
    # SQL
    if re.search(r'\b(SELECT|INSERT|UPDATE|DELETE|CREATE\s+TABLE|DROP|ALTER)\b', content, re.IGNORECASE):
        return 'sql'
    
    # Go
    if re.search(r'\bfunc\s+\w+\(|\btype\s+\w+\s+(struct|interface)|package\s+\w+', content):
        return 'go'
    
    # TypeScript/JavaScript
    if re.search(r'\b(import|export|const|let|var|function|class|interface|type)\s', content):
        return 'typescript'
    
    # INI format
    if re.search(r'^\[.*\]$', content, re.MULTILINE) and re.search(r'^\w+=', content, re.MULTILINE):
        return 'ini'
    
    # Bash/Shell (environment variables, commands)
    if re.search(r'\b(sudo|apt|yum|docker|npm|yarn|git|chmod|chown|mkdir|cd|ls|cat|grep|sed|awk|echo|export|source)\s', content):
        return 'bash'
    if re.search(r'^\w+=[^\s]', content, re.MULTILINE) and not content.strip().startswith('<'):
        return 'bash'
    
    # Check context clues
    if 'response' in prev_text or 'output' in prev_text or 'example' in prev_text:
        # Check if it's simple text output
        if re.search(r'http://|https://', content) and len(content.strip().split('\n')) <= 3:
            return 'text'
        if content.strip().isdigit():
            return 'text'
        if re.search(r'^[\w\s\.\-:]+$', content.strip()):
            return 'text'
    
    # Headers (HTTP)
    if 'header' in prev_text and re.search(r'^[\w-]+:\s*.+$', content, re.MULTILINE):
        return 'http'
    
    # Mermaid diagrams
    if re.search(r'(sequenceDiagram|graph|flowchart|classDiagram)', content):
        return 'mermaid'
    
    # Default to text for short simple content
    if len(content.strip().split('\n')) <= 2 and not re.search(r'[{}\[\]<>]', content):
        return 'text'
    
    return 'text'

def fix_file(filepath):
    """Fix all code blocks without language tags in a file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    result = []
    i = 0
    changes = 0
    
    while i < len(lines):
        if lines[i].strip() == '```':
            # Found a code block without language
            prev_lines = [lines[j].strip() for j in range(max(0, i-5), i) if lines[j].strip()]
            
            # Collect content
            content_lines = []
            i += 1
            start_i = i
            while i < len(lines) and not lines[i].strip().startswith('```'):
                content_lines.append(lines[i])
                i += 1
            
            # Collect next lines for context
            next_lines = [lines[j].strip() for j in range(i+1, min(i+4, len(lines))) if j < len(lines) and lines[j].strip()]
            
            content = ''.join(content_lines)
            lang = determine_language(content, prev_lines, next_lines)
            
            # Write fixed version
            result.append(f'```{lang}\n')
            result.extend(content_lines)
            if i < len(lines):
                result.append(lines[i])  # closing ```
            changes += 1
        else:
            result.append(lines[i])
        i += 1
    
    if changes > 0:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.writelines(result)
        return changes
    return 0

def main():
    if len(sys.argv) < 2:
        print("Usage: fix_codeblocks.py <file1> [file2] ...")
        sys.exit(1)
    
    total = 0
    for filepath in sys.argv[1:]:
        path = Path(filepath)
        if not path.exists():
            print(f"File not found: {filepath}")
            continue
        
        changes = fix_file(filepath)
        if changes > 0:
            print(f"✓ Fixed {changes} code blocks in {path.name}")
            total += changes
        else:
            print(f"  No changes needed in {path.name}")
    
    print(f"\n{'='*50}")
    print(f"Total: Fixed {total} code blocks across {len(sys.argv)-1} files")

if __name__ == '__main__':
    main()

File: fix_codeblocks.js
#!/usr/bin/env bun
/**
 * Automatically fix code blocks missing language tags in MDX files.
 */

const fs = require('fs');
const path = require('path');

function determineLanguage(content, prevLines, nextLines) {
    const contentLower = content.toLowerCase();
    const prevText = prevLines.slice(-5).join('\n').toLowerCase();
    const nextText = nextLines.slice(0, 3).join('\n').toLowerCase();
    
    // HTTP endpoints
    if (/^(GET|POST|PUT|DELETE|PATCH|HEAD|OPTIONS)\s+\//m.test(content)) {
        return 'http';
    }
    
    // curl commands
    if (/^\s*curl\s/m.test(content)) {
        return 'bash';
    }
    
    // YAML/Docker Compose
    if (/^(services:|version:|networks:|volumes:|image:|ports:|environment:)/m.test(content)) {
        return 'yaml';
    }
    
    // Caddyfile
    if (/^\w+\.(com|dev|local|net)\s*{/m.test(content) || prevText.includes('caddy')) {
        return 'caddy';
    }
    
    // JSON
    if ((content.trim().startsWith('{') || content.trim().startsWith('[')) && /["']:\s*["'{[]/m.test(content)) {
        return 'json';
    }
    
    // XML
    if (/<\?xml|<roblox|<Item|<Properties|<\/\w+>/m.test(content)) {
        return 'xml';
    }
    
    // Lua
    if (/(function|local|if\s+.+\s+then|end\b|return\b)/m.test(content)) {
        return 'lua';
    }
    
    // Dockerfile
    if (/^(FROM|RUN|COPY|WORKDIR|ENTRYPOINT|CMD|EXPOSE|ENV|ARG)\s/m.test(content)) {
        return 'dockerfile';
    }
    
    // SQL
    if (/\b(SELECT|INSERT|UPDATE|DELETE|CREATE\s+TABLE|DROP|ALTER)\b/im.test(content)) {
        return 'sql';
    }
    
    // Go
    if (/\bfunc\s+\w+\(|\btype\s+\w+\s+(struct|interface)|package\s+\w+/.test(content)) {
        return 'go';
    }
    
    // TypeScript/JavaScript
    if (/\b(import|export|const|let|var|function|class|interface|type)\s/.test(content)) {
        return 'typescript';
    }
    
    // INI format
    if (/^\[.*\]$/m.test(content) && /^\w+=/m.test(content)) {
        return 'ini';
    }
    
    // Bash/Shell
    if (/\b(sudo|apt|yum|docker|npm|yarn|git|chmod|chown|mkdir|cd|ls|cat|grep|sed|awk|echo|export|source)\s/.test(content)) {
        return 'bash';
    }
    if (/^\w+=[^\s]/m.test(content) && !content.trim().startsWith('<')) {
        return 'bash';
    }
    
    // Context clues for responses/outputs
    if (prevText.includes('response') || prevText.includes('output') || prevText.includes('example')) {
        if (/https?:\/\//.test(content) && content.trim().split('\n').length <= 3) {
            return 'text';
        }
        if (/^\d+$/.test(content.trim())) {
            return 'text';
        }
        if (/^[\w\s\.\-:]+$/.test(content.trim())) {
            return 'text';
        }
    }
    
    // Headers (HTTP)
    if (prevText.includes('header') && /^[\w-]+:\s*.+$/m.test(content)) {
        return 'http';
    }
    
    // Mermaid diagrams
    if (/(sequenceDiagram|graph|flowchart|classDiagram)/.test(content)) {
        return 'mermaid';
    }
    
    // Default to text for short simple content
    if (content.trim().split('\n').length <= 2 && !/[{}\[\]<>]/.test(content)) {
        return 'text';
    }
    
    return 'text';
}

function fixFile(filepath) {
    const content = fs.readFileSync(filepath, 'utf-8');
    const lines = content.split('\n').map(line => line + '\n');
    // Handle last line without newline
    if (lines[lines.length - 1] === '\n') {
        lines[lines.length - 1] = '';
    }
    
    const result = [];
    let i = 0;
    let changes = 0;
    
    while (i < lines.length) {
        if (lines[i].trim() === '```') {
            // Found code block without language
            const prevLines = [];
            for (let j = Math.max(0, i - 5); j < i; j++) {
                if (lines[j].trim()) {
                    prevLines.push(lines[j].trim());
                }
            }
            
            // Collect content
            const contentLines = [];
            i++;
            while (i < lines.length && !lines[i].trim().startsWith('```')) {
                contentLines.push(lines[i]);
                i++;
            }
            
            // Collect next lines for context
            const nextLines = [];
            for (let j = i + 1; j < Math.min(i + 4, lines.length); j++) {
                if (lines[j].trim()) {
                    nextLines.push(lines[j].trim());
                }
            }
            
            const blockContent = contentLines.join('');
            const lang = determineLanguage(blockContent, prevLines, nextLines);
            
            // Write fixed version
            result.push(`\`\`\`${lang}\n`);
            result.push(...contentLines);
            if (i < lines.length) {
                result.push(lines[i]); // closing ```
            }
            changes++;
        } else {
            result.push(lines[i]);
        }
        i++;
    }
    
    if (changes > 0) {
        fs.writeFileSync(filepath, result.join(''), 'utf-8');
        return changes;
    }
    return 0;
}

function main() {
    const args = process.argv.slice(2);
    
    if (args.length === 0) {
        console.log('Usage: fix_codeblocks.js <file1> [file2] ...');
        process.exit(1);
    }
    
    let total = 0;
    for (const filepath of args) {
        if (!fs.existsSync(filepath)) {
            console.log(`File not found: ${filepath}`);
            continue;
        }
        
        const changes = fixFile(filepath);
        const filename = path.basename(filepath);
        if (changes > 0) {
            console.log(`✓ Fixed ${changes} code blocks in ${filename}`);
            total += changes;
        } else {
            console.log(`  No changes needed in ${filename}`);
        }
    }
    
    console.log('\n' + '='.repeat(50));
    console.log(`Total: Fixed ${total} code blocks across ${args.length} files`);
}

main();
